backend/src/users/services/profile_service.py

```diff
@@ -0,0 +1,194 @@
+
+
+class ProfileService:
+    def __init__(
+        self,
+        repository: ProfileRepository,
+        cloud_storage: CloudStorage,
+        db: Session,
+    ):
+        self.repository = repository
+        self.cloud_storage = cloud_storage
+        self.db = db
+        self.start_listening_threads()
+
+    def start_listening_threads(self):
+        run_handler_thread(
+            cls=self,
+            subscribe_name="user_created",
+            handler_name="event_handler_user_created",
+        )
+        run_handler_thread(
+            cls=self,
+            subscribe_name="user_deleted",
+            handler_name="event_handler_user_deleted",
+        )
+
+    def event_handler_user_created(
+        self,
+        cache_handler: Optional[CacheHandler],
+        subscribe_name: Optional[str],
+    ):
+        cache_handler.subscribe_event(subscribe_name)
+        while True:
+            time.sleep(0.2)
+            data = cache_handler.get_event()
+            if data:
+                print(data)
+                user_id = UUID(data["id"])
+                profile = data["profile"]
+                self.create_related_profile(
+                    user_id=user_id,
+                    profile=profile,
+                    db=self.db,
+                )
+                cache_handler.unsubscribe_event("user_created")
+
+    def event_handler_user_deleted(
+        self,
+        cache_handler: Optional[CacheHandler],
+        subscribe_name: Optional[str],
+    ):
+        cache_handler.subscribe_event(subscribe_name)
+        while True:
+            time.sleep(0.2)
+            data = cache_handler.get_event()
+            if data:
+                print(data)
+                user_id = UUID(data["id"])
+                self.delete_related_profile(
+                    user_id=user_id,
+                )
+                cache_handler.unsubscribe_event("user_deleted")
+
+    def create_related_profile(
+        self,
+        user_id: UUID,
+        profile: CreateProfileModel,
+        db: Session,
+    ):
+        try:
+            # Create profile in the repository
+            profile_db = self.repository.create_profile(
+                user_id=user_id,
+                name=profile.get("name"),
+                surname=profile.get("surname"),
+                db=db,
+            )
+            if profile_db:
+                # Prepare and upload profile image to cloud storage
+                image = prepare_profile_image(
+                    image_url=profile.get("image_url"),
+                    user_id=user_id,
+                )
+
+                # Upload profile image to cloud storage
+                self.cloud_storage.upload_file(
+                    **image,
+                )
+            else:
+                raise InvalidProfileData
+
+        except Exception:
+            raise ProfileCreationFailed
+
+    @staticmethod
+    def delete_related_profile(user_id: UUID):
```

Comment:
I think this method is misleading. What is the purpose of it ?

Reply:
Generally, the user creates a user model and then sends pubsub to the profile to create the profile, but in the profile we create a mini avatar, so to keep it logical,
when we do a similar thing when deleting, by deleting the user model via the sqlalchemy relation, we delete the profile model automatically, but we still have to delete the avatar and it performs this process.